File: 2-NORMAL/Python/1-binary-search-tree-operations/proposed-solution/main.py
# Challenge: Implement a binary search tree (BST) that supports insertions, searches, and deletions.
# Test different traversals (in-order, pre-order, and post-order) to ensure the tree updates correctly.

from bst import BinarySearchTree

def main():
    bst = BinarySearchTree()
    # Insert nodes
    for key in [50, 30, 20, 40, 70, 60, 80]:
        bst.insert(key)

    print("Inorder traversal:")
    print(bst.inorder())

    print("Preorder traversal:")
    print(bst.preorder())

    print("Postorder traversal:")
    print(bst.postorder())

    # Search for a key
    key = 40
    found = bst.search(key)
    print(f"Search for {key}: {'Found' if found else 'Not found'}")

    # Delete a key
    bst.delete(20)
    print("Inorder traversal after deleting 20:")
    print(bst.inorder())

    bst.delete(30)
    print("Inorder traversal after deleting 30:")
    print(bst.inorder())

    bst.delete(50)
    print("Inorder traversal after deleting 50:")
    print(bst.inorder())

if __name__ == "__main__":
    main()

File: 2-NORMAL/JavaScript/1-binary-search-tree-operations/proposed-solution/main.js
/*
Challenge: Implement a binary search tree (BST) that supports insertions, searches, and deletions.
Test different traversals (in-order, pre-order, and post-order) to ensure the tree updates correctly.
*/

// main.js - Example usage of BinarySearchTree

import { BinarySearchTree } from './bst.js';

function main() {
  const bst = new BinarySearchTree();
  [50, 30, 20, 40, 70, 60, 80].forEach((key) => bst.insert(key));

  console.log("Inorder traversal:");
  console.log(bst.inorder());

  console.log("Preorder traversal:");
  console.log(bst.preorder());

  console.log("Postorder traversal:");
  console.log(bst.postorder());

  const key = 40;
  const found = bst.search(key);
  console.log(`Search for ${key}: ${found ? "Found" : "Not found"}`);

  bst.delete(20);
  console.log("Inorder traversal after deleting 20:");
  console.log(bst.inorder());

  bst.delete(30);
  console.log("Inorder traversal after deleting 30:");
  console.log(bst.inorder());

  bst.delete(50);
  console.log("Inorder traversal after deleting 50:");
  console.log(bst.inorder());
}

main();
